Jenkinsfile

```diff
@@ -0,0 +1,139 @@
+import com.esri.zrh.jenkins.ce.CityEnginePipelineLibrary
+import com.esri.zrh.jenkins.ce.PrtAppPipelineLibrary
+import groovy.transform.Field
+
+@Field def psl = PslFactory.create(this, UploadTrackingPsl.ID)
+@Field def cepl = new CityEnginePipelineLibrary(this, psl)
+@Field def papl = new PrtAppPipelineLibrary(cepl)
+
+
+// -- PIPELINE SETUP
+
+psl.runsHere('testing')
+env.PIPELINE_ARCHIVING_ALLOWED = "true"
+properties([ disableConcurrentBuilds() ])
+
+
+// -- CONFIGURATION
+
+@Field final String REPO = 'git@github.com:esri/puma.git'
+@Field final String SOURCE = 'puma.git'
+@Field final String SOURCE_STASH = 'puma-sources'
+
+@Field final String DOCKER_AGENT_WINDOWS = 'win19-64-d'
+@Field final String DOCKER_WS_WINDOWS = "c:/temp/puma/ws"
+
+@Field final Map WINDOWS_DOCKER_CONFIG = [ ba: DOCKER_AGENT_WINDOWS, ws: DOCKER_WS_WINDOWS ]
+@Field final Map WINDOWS_NATIVE_CONFIG = [ os: cepl.CFG_OS_WIN10, bc: cepl.CFG_BC_REL, tc: cepl.CFG_TC_VC1427, cc: cepl.CFG_CC_OPT, arch: cepl.CFG_ARCH_X86_64 ]
+
+@Field final Map RHINO6_CONFIG = [ rh: '6.35.21222.17001', rhsdk: '6.35.21222.17001' ]
+@Field final Map RHINO7_CONFIG = [ rh: '7.17.22102.5001',  rhsdk: '7.19.22165.13001' ]
+
+@Field final List CONFIGS_PREPARE = [
+    WINDOWS_NATIVE_CONFIG
+]
+
+@Field final List CONFIGS = [
+	composeConfig(RHINO6_CONFIG, WINDOWS_NATIVE_CONFIG, WINDOWS_DOCKER_CONFIG),
+	composeConfig(RHINO7_CONFIG, WINDOWS_NATIVE_CONFIG, WINDOWS_DOCKER_CONFIG),
+]
+
+
+// -- THE PIPELINE
+
+stage('prepare') {
+	cepl.runParallel(taskGenPrepare())
+}
+
+stage('build') {
+	cepl.runParallel(taskGenPuma())
+}
+
+
+// -- TASK GENERATORS
+
+Map taskGenPrepare() {
+	Map tasks = [:]
+	tasks << cepl.generateTasks('prepare', this.&taskPrepare, CONFIGS_PREPARE)
+	return tasks
+}
+
+Map taskGenPuma() {
+	return cepl.generateTasks('build', this.&taskBuildPuma, CONFIGS)
+}
+
+
+// -- TASK IMPLEMENTATIONS
+
+def taskPrepare(cfg) {
+ 	cepl.cleanCurrentDir()
+	papl.checkout(REPO, env.BRANCH_NAME)
+	stash(name: SOURCE_STASH)
+}
+
+def taskBuildPuma(cfg) {
+	cepl.cleanCurrentDir()
+	unstash(name: SOURCE_STASH)
+
+	final String tag = "rh${cfg.rh}-rhsdk${cfg.rhsdk}"
+    final String image = "zrh-dreg-sp-1.esri.com/puma/puma-toolchain:${tag}"
+
+	final String buildCmd = "ci_build.cmd"
+	final String containerName = "puma-build-rh${cfg.rh}-${env.BRANCH_NAME.replaceAll('/', '_')}-b${BUILD_ID}"
+
+	String workDir = "${cfg.ws}/${SOURCE}"
+	Map dirMap = [ (env.WORKSPACE) : cfg.ws ]
+
+	final String rhinoMajorVersion = cfg.rh.tokenize('.')[0]
+	Map envMap = [ 'RHINO_VER_MAJOR' : rhinoMajorVersion ]
+
+	runDockerCmd(cfg, image, containerName, dirMap, envMap, workDir, buildCmd)
+
+	manifest = readYaml(file: "${SOURCE}/manifest.yml") // called within publish, different cwd
+
+	// build final package version as <major>.<minor>.<build number>
+	final List pkgVerComponents = manifest.version.tokenize('.')
+	pkgVerComponents[2] = BUILD_ID
+	final String pkgVer = pkgVerComponents.join('.')

```

Comment:
currently builds are labelled as MAJOR.MINOR.BUILDNR, maybe label them in such a way that it's obvious that the third number corresponds to the build number. (Analogous to other plugins)

Reply:
deferring switchting to semantic versioning to a separate ticket as it affects also how we display the version in the UI (versus rhino only allows the a.b.c.d syntax)